dkms.in

```diff
@@ -833,6 +816,90 @@ run_build_script() {
     fi
 }
 
+# helper functions for getting cc/ld used by kernel
+detect_toolchain_version() {
+    # $1 = program name, e.g. gcc, ld.lld
+    # $2 = major version
+    # $3 = minor version
+    # $4 = patch version, optional
+    local -a filenames
+    if [[ "$4" ]]; then
+        filenames+=( "$1-$(printf "%d.%d.%d" "$2" "$3" "$4")" )
+    fi
+    filenames+=(
+        "$1-$(printf "%d.%d" "$2" "$3")"
+        "$1-$(printf "%d" "$2")"
+        "$1"
+    )
+    for f in "${filenames[@]}"; do
+        if command -V "$f" &>/dev/null; then
+            echo "$f"
+            return 0
+        fi
+    done
+    return 1
+}
+detect_toolchain_from_text() {
+    # $1 = program version text, contains x.y.z
+    # $2 = program name, e.g. gcc, clang
+    local patterns=(
+        's/^.*\b([0-9]+\.[0-9]+\.[0-9]+).*$/\1/p'
+    )
+    local -a ver
+    for p in "${patterns[@]}"; do
+        readarray -d '.' -t ver <<< "$(echo "$1" | sed -En "$p")"
+        if [[ "$ver" ]]; then
+            if detect_toolchain_version "$2" ${ver[0]} ${ver[1]} ${ver[2]}; then
+                return 0
+            fi
+        fi
+    done
+    return 1
+}
+detect_toolchain_from_kconfig() {
+    # $1 = program name in kconfig, e.g. GCC, CLANG, LD, LLD
+    # $2 = program name, e.g. gcc, clang, ld.bfd, ld.lld
+    local ver_full="$(grep "^CONFIG_$1_VERSION=" "${kernel_config}" | cut -f2 -d= )"
+    if [[ -z "${ver_full}" ]] || [[ "${ver_full}" = 0 ]]; then
+        return 1
+    fi
+    local ver_major="${ver_full:0: -4}"
```

Comment:
Let's add a couple of examples for illustration purposes - one GCC and one CLANG. For example my system has `CONFIG_GCC_VERSION=60300`
Looking at the code below - seems like ver_patch is always set so one can drop the explicit check `if [[ "$4" ]]` in detect_toolchain_version, right?